winml/lib/Api.Ort/OnnxruntimeEngineBuilder.cpp

```diff
@@ -41,10 +41,9 @@ STDMETHODIMP OnnxruntimeEngineBuilder::CreateEngine(_winml::IEngine** out) {
     RETURN_HR_IF_NOT_OK_MSG(ort_api->AddFreeDimensionOverride(session_options.get(), DATA_BATCH, batch_size_override_.value()),
                             ort_api);
   }
-
-  for (const auto& [name, value] : named_dimension_overrides_) {
-    std::string narrow_name = std::wstring_convert<std::codecvt_utf8<wchar_t>>().to_bytes(name.c_str());
-    ort_api->AddFreeDimensionOverrideByName(session_options.get(), narrow_name.c_str(), value);
+  for (auto&& override : named_dimension_overrides_) {
```

Comment:
Was there a problem with `const auto&`?

Reply:
Added back the const, but auto&& should be a bit more performant by avoiding copying the KeyValue pairs since it will use an r-value. Honestly I doubt it is that relevant a difference but I don't see why not.

Reply:
const auto& never copies, and you can take const references to rvalues.

Reply:
A bit more in depth:

References won't copy. const lvalue references work everywhere (also for rvalues, extending the lifetime of the rvalue until the reference dies). For example (adapted from cppreference):

```
std::string s1 = "Test"; 
f(s1 + s1);  // call some function with s1 + s1. Here, s1 + s1 is a temporary rvalue, and the temporary string will be destroyed after calling f


{
    const std::string& r2 = s1 + s1;  // here, s1 + s1 is a temporary rvalue. You can assign it to a lvaue reference. Since you've assigned the result to a reference, it's not destroyed until the reference goes out of scope
    // ...
    // some more commands
    // ...
}  // end of the scope that contained const std::string& r2 -> string is destroyed
```

const auto& is what you want, unless you need to modify the items.

If you have int&&, it's a rvalue reference. These are usually used in function signatures if you want an overload that takes an rvalue (to e.g. avoid copies).

auto&& or T&&, where T is a template parameter, is called a forwarding reference or universal reference and it keeps the "valueness" of the thing it's assigned. So if you initialize it with an lvalue it will be an lvalue ref, with an rvalue it will be an rvalue ref (works everywhere). You usually use it when you want to change the thing (so can't be const&) and want it to work with both rvalue and lvalue (which is useful in templated code, but can be used elsewhere).

Reply:
const auto& should suffice

Reply:
thanks for the explanation